ui: add button labels that tell the user which button does what

hnez opened the pull request:
This adds a little legend on the right of the screen (where the two buttons are) that tells the user what which button does what:

<img alt="dig_out" src="https://github.com/linux-automation/tacd/assets/1273320/d3c7d96c-3c0d-413e-a15c-539208f4f411" width="23%">
<img alt="iobus" src="https://github.com/linux-automation/tacd/assets/1273320/a2e37213-824e-4693-b566-f67e03b1ed2c" width="23%">
<img alt="iobus_health" src="https://github.com/linux-automation/tacd/assets/1273320/e0750b5b-356c-4be4-8bde-b3f73e0face7" width="23%">
<img alt="locator" src="https://github.com/linux-automation/tacd/assets/1273320/1db44902-6b53-4f30-86f8-4690c1634821" width="23%">
<img alt="overtemperature" src="https://github.com/linux-automation/tacd/assets/1273320/fab8de49-330c-4ef8-87f5-42992d4757d4" width="23%">
<img alt="power" src="https://github.com/linux-automation/tacd/assets/1273320/6a5907a5-13cc-467a-bca7-2f35eef61306" width="23%">
<img alt="power_fail" src="https://github.com/linux-automation/tacd/assets/1273320/2d4ef94c-4cbb-4ba3-a467-834a3586cfd7" width="23%">
<img alt="reboot" src="https://github.com/linux-automation/tacd/assets/1273320/9c3f1cd8-72ce-4ce6-81b9-a4ae63c956c0" width="23%">
<img alt="screensaver" src="https://github.com/linux-automation/tacd/assets/1273320/cbb9cf6a-a19d-4538-bf31-810e9d4f6072" width="23%">
<img alt="system" src="https://github.com/linux-automation/tacd/assets/1273320/852f5977-0c3a-4393-a17a-e589125810fe" width="23%">
<img alt="uart" src="https://github.com/linux-automation/tacd/assets/1273320/70f2a03e-1664-4d7a-85f6-a200b49b43f5" width="23%">
<img alt="update_available" src="https://github.com/linux-automation/tacd/assets/1273320/1363c658-3676-479e-b64f-565efe104a7b" width="23%">
<img alt="update_installing" src="https://github.com/linux-automation/tacd/assets/1273320/3899198c-6d31-4294-bbed-a06104dc4318" width="23%">
<img alt="usb" src="https://github.com/linux-automation/tacd/assets/1273320/4e4e2813-70e2-4449-a508-83713de71f08" width="23%">
<img alt="usb_overload" src="https://github.com/linux-automation/tacd/assets/1273320/d898f6bd-65cf-4c6c-b755-348f2e89b6ee" width="23%">
<img alt="reboot_slot" src="https://github.com/linux-automation/tacd/assets/1273320/53c62834-6b2e-4e6c-a305-85c97a35c90d" width="23%">


TODO before merging:

  - [x] Add comments to the drawing routine that make it easier to read
  - [x] Add comments to the screen rotation trickery
  - [x] Add legends to alert screens as well

hnez commented on src/ui/display.rs:
```diff
@@ -176,3 +186,28 @@ impl OriginDimensions for DisplayExclusive {
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        let res = self.size();
```
Add a comment about this only doing one kind of rotation (the one we need for the button legend).

hnez commented on src/ui/display.rs:
```diff
@@ -176,3 +186,28 @@ impl OriginDimensions for DisplayExclusive {
+
+impl<'a> OriginDimensions for DisplayRotated<'a> {
+    fn size(&self) -> Size {
+        self.inner.size()
```
Our display is square, but this should flip the x/y coordinates anyways.

hnez commented on src/ui/screens/system.rs:
```diff
@@ -93,8 +95,8 @@ impl ActivatableScreen for SystemScreen {
                 Box::new(|info: &LinkInfo| match info.carrier {
-                    true => format!("Uplink: {}MBit/s", info.speed),
-                    false => "Uplink: Down".to_string(),
+                    true => format!("UL:  {}MBit/s", info.speed),
+                    false => "UL:  Down".to_string(),
```
Split this into a separate commit and write about the _why_ in the commit message (the IP address line got too wide with the button legend so it had to move left without breaking alignment of the rows, so "Uplink" had to become shorter).

hnez commented on src/ui/screens/system.rs:
```diff
@@ -75,6 +75,8 @@ impl ActivatableScreen for SystemScreen {
     fn activate(&mut self, ui: &Ui, display: Display) -> Box<dyn ActiveScreen> {
         draw_border("System Status", SCREEN_TYPE, &display);
 
+        display.with_lock(|target| draw_button_legend(target, "Next / Do", "Screen"));
+
```
`draw_border()` followed by `with_lock(|t| draw_button_legend)` seems to be a common pattern now. Check if we should make `draw_border` accept a `DisplayExclusive` like `draw_button_legend` does, so both can be drawn in the same `.with_lock()` call.

hnez commented on src/ui/widgets.rs:
```diff
@@ -98,6 +99,51 @@ pub struct DynamicWidget<T: Sync + Send + 'static> {
+pub fn draw_button_legend(target: &mut DisplayExclusive, lower: &str, upper: &str) -> Rectangle {
+    let mut target = target.rotate();
+
+    let ui_text_style: MonoTextStyle<BinaryColor> =
+        MonoTextStyle::new(&SMALL_TEXT_FONT, BinaryColor::On);
+
+    Text::with_alignment(lower, Point::new(73, 236), ui_text_style, Alignment::Center)
+        .draw(&mut target)
+        .unwrap();
+
+    Text::with_alignment(
+        upper,
+        Point::new(168, 236),
+        ui_text_style,
+        Alignment::Center,
+    )
+    .draw(&mut target)
+    .unwrap();
+
+    Line::new(Point::new(27, 224), Point::new(213, 224))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
+
+    Line::new(Point::new(120, 224), Point::new(120, 240))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
+
+    Line::new(Point::new(26, 225), Point::new(26, 240))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
+
+    Line::new(Point::new(214, 225), Point::new(214, 240))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
```
Add a comment for each draw call.

hnez commented on src/ui/widgets.rs:
```diff
@@ -248,6 +294,22 @@ impl<T: Serialize + DeserializeOwned + Send + Sync + Clone + 'static> DynamicWid
         )
     }
 
+    pub fn button_legend(
```
Add a docstring.

hnez commented:
I've now gone through all screens and added legends.
With that I am ready for review.

SmithChart commented:
First: I find this really helpful. It feels like the first time that I really know what button does which action. Or at least this way a user can learn to use them right. (Maybe we can add caps with different colors to the buttons, to make it even easier to remember?)

@fiurin  and I had a look at this change from a user-perspective. One thing we kind of stumbled across is the label on the lower button. This button can either be used with a short- or long press. And does different things depending on the type of press.
For screens with more than one option the `toggle` or `select` labels are not clear. We would vote to use `Action` here instead. This is intended to be ominous and vague, so the user can assume that both long- and short press are available here. This applies to the "USB Host", "Digital out", "DUT UART" and "System screens".

I would vote to leave all other screens as they are in this PR.

hnez commented:
I've rebased the PR on top of recent main (it should go in after #59 though, so we see the CI succeed before merging) and changed the labels in the mentioned screens from what they were there to "Action".

SmithChart commented:
> I've rebased the PR on top of recent main (it should go in after #59 though, so we see the CI succeed before merging) and changed the labels in the mentioned screens from what they were there to "Action".

Thanks! I tried the recent changes on my device. Looks good!

hnez commented:
Using an actual graphics toolkit, like LVGL for instance, would be great I think, because it could make the UI more modern (and maybe make working on the UI a bit easier).
I will propose an evaluation of GUI options for the TAC and other products with tiny displays and reduced input options as a techweek project once that comes around.
